Reduce memory usage

PotatoesFall opened the pull request:
Hi!

Hope it's cool if I made a PR, I really liked this tool but I noticed it keeps entropy information about all lines for a given file in memory and does a huge sort at the end.

## Proposed change

I propose this change, which means we never keep more lines in memory than necessary.

The `Entropies` struct keeps the top n lines sorted in a slice.

Testing with a medium sized repository, I noticed the old version got all the way up to 1G memory consumption, after the change it doesn't even show up in my top 100 memory consumption programs.

## Execution time

While I didn't see an improvement in execution time for my testing, this version does get rid of the large sort towards the end.

If we want to cut down execution time in the future, it may be wise to make an individual list per file and/or per directory (as before), and merge these together when done with each file/directory, to reduce locking on the main `Entropies` struct.

For very large values of `-top`, we could possibly also get very small performance gains using a max heap vs a slice, but that's probably premature optimization at this point.

PotatoesFall committed:
reduce memory usage

PotatoesFall committed:
make memory safe

EwenQuim commented:
Thanks for your contribution!

I'll look at it tomorrow ;)

AlexanderYastrebov commented on main.go:
```diff
@@ -35,6 +34,36 @@ type Entropy struct {
+	es.Lock()
+	defer es.Unlock()
+
+	entropies := es.Entropies
+	if entropies[len(entropies)-1].Entropy >= e.Entropy {
+		return
+	}
+
+	for i := range len(entropies) {
+		if entropies[i].Entropy < e.Entropy {
+			for j := len(entropies) - 1; j > i; j-- {
+				entropies[j] = entropies[j-1]
+			}
+			entropies[i] = e
+			return
+		}
+	}
```
```go
i, _ := slices.BinarySearchFunc(es.Entropies, e, func(a, b Entropy) int {
	return b.Entropy - a.Entropy
})

if i < len(es.Entropies) {
	copy(es.Entropies[i+1:], es.Entropies[i:])
	es.Entropies[i] = e
}
```

AlexanderYastrebov replied:
Also
```go
func entropy(text string) float64 {
	unique := make(map[byte]float64, len(text))
	for i := 0; i < len(text); i++ { // count bytes, its faster than counting runes
		unique[text[i]] += 1.0
	}
	entropy := 0.0
	for _, count := range unique {
		res := count / float64(len(text)) // res can not be 0 because count and len are not zero
		entropy -= res * math.Log2(res)
	}
	return entropy
}
```

PotatoesFall replied:
Nice, thanks!

### binary search

I added a modified version of the first suggestion. I left in the initial check for `if entropies[len(entropies)-1].Entropy >= e.Entropy { ...` since in most cases, the majority of lines won't need to be added at all, and we can potentially skip unnecessary O(logN) operations with this check.

Also needed to modify the way float difference is returned as an int.

### entropy algorithm

looks very good, but IMO not suited for this MR

PotatoesFall committed:
use binary search for entropy tracking

AlexanderYastrebov commented on main.go:
```diff
@@ -35,6 +35,38 @@ type Entropy struct {
+		return 0
+	})
+
+	copy(es.Entropies[i+1:], es.Entropies[i:])
+	es.Entropies[i] = e
```
I think we need `if i < len(es.Entropies)` check here as added item may evict the last one (i == len(es.Entropies))

PotatoesFall replied:
We don't need this, when we are evicting the last element, the `copy` simply becomes a no-op.

Evicting the last item is `i == len(es.Entropies)-1`, not `i == len(es.Entropies)`

lpmi-13 committed:
reorder build steps to improve layer caching

Currently, any change in the source code requires a rerun of `go mod download`. This commit moves that step earlier so that only changes in the dependencies require a rebuild of that layer.

EwenQuim commented:
Can you please resolve conflicts? Thanks!

PotatoesFall committed:
Merge remote-tracking branch 'origin/master' into memory

EwenQuim commented:
That is some very high quality PR @PotatoesFall, thank you!

Some tests needed to be updated though, I did it and merged it.
